src/components/hypercube-table.jsx

```diff
@@ -78,133 +88,129 @@ function noRowsRenderer() {
   return <div className="no-values">No values</div>;
 }
 
-const GLYPH_SIZE = 6;
-
+const GLYPH_SIZE = 8;
+const MAX_ROW_GLYTH_LENGTH = 40;
 function getMeasureWidth(layout, measureIndex, measureName) {
-  const dataSize = layout.qHyperCube.qMeasureInfo.length > 0 ? layout.qHyperCube.qMeasureInfo[measureIndex].qApprMaxGlyphCount : 0;
-  const titleSize = measureName.length;
-  const size = Math.max(dataSize, titleSize);
-  return size * GLYPH_SIZE;
+  const dataSize = layout.qHyperCube.qMeasureInfo.length > measureIndex ? layout.qHyperCube.qMeasureInfo[measureIndex].qApprMaxGlyphCount : 0;
+  const titleSize = (measureName || '').length;
+  const size = Math.min(Math.max(dataSize, titleSize), MAX_ROW_GLYTH_LENGTH);
+  return size * GLYPH_SIZE + 8;
 }
 
 function getDimensionWidth(layout, dimensionIndex, dimensionName) {
   const dataSize = layout.qHyperCube.qDimensionInfo.length > dimensionIndex ? layout.qHyperCube.qDimensionInfo[dimensionIndex].qApprMaxGlyphCount : 0;
-  const titleSize = dimensionName.length;
-  const size = Math.max(dataSize, titleSize);
-  return size * GLYPH_SIZE;
+  const titleSize = (dimensionName || '').length;
+  const size = Math.min(Math.max(dataSize, titleSize), MAX_ROW_GLYTH_LENGTH);
+  return size * GLYPH_SIZE + 8;
 }
 
-export class HypercubeTable extends React.Component {
-  constructor() {
-    super();
-    this.state = {};
-    this.selfRef = React.createRef();
+function getTotalWidth(layout, dimensions, measures) {
+  if (!layout) {
+    return 0;
   }
+  const dimSizes = dimensions.map((dim, dimensionIndex) => getDimensionWidth(layout, dimensionIndex, dim.title));
+  const mesSizes = measures.map((measure, measureIndex) => getMeasureWidth(layout, measureIndex, measure.title));
+  return 20 + dimSizes.reduce((a, b) => a + b, 0) + mesSizes.reduce((a, b) => a + b, 0);
+}
 
-  async componentWillReceiveProps() {
-    const { app, measure, dimensions } = this.props;
-
-    const hypercubeProps = {
-      qInfo: {
-        qId: 'measurebox1',
-        qType: 'measurebox1',
-      },
-      qHyperCubeDef: {
-        qMeasures: [{
-          qLibraryId: measure.qInfo.qId,
-        }],
-        qInitialDataFetch: [
-          {
-            qTop: 0,
-            qLeft: 0,
-            qHeight: 50,
-            qWidth: dimensions.length + 1,
-          },
-        ],
-      },
-    };
-
-    if (dimensions.length > 0) {
-      hypercubeProps.qHyperCubeDef.qDimensions = dimensions.map(fieldName => ({ qDef: { qFieldDefs: [fieldName] } }));
-    } else {
-      hypercubeProps.qHyperCubeDef.qDimensions = [];
-    }
-
-    if (!this.object) {
-      this.object = await app.createSessionObject(hypercubeProps);
-    } else {
-      await this.object.setProperties(hypercubeProps);
-    }
-    const layout = await this.object.getLayout();
-    this.setState({ layout, object: this.object });
+export default function HypercubeTable({
+  app, measures, dimensions, height, maxWidth, onHeaderClick,
+}) {
+  const hypercubeProps = {
```

Comment:
Move this out to a separate function (like `createDefinition()`), easier to test in the future.